### PR TITLE
ecs_ecr - Add encryption_configuration option

### LINKED ISSUE #1203: ecs_ecr - support for specifying KMS key
### Summary

Unless I'm missing something, I don't see a way to specify KMS key usage in this module.  Is it possible to integrate this? 

### Issue Type

Feature Idea

### Component Name

ecs_ecr

### Additional Information

```yaml (paste below)

```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by rwha**
##### SUMMARY
Adds an encryption_configuration option for new repositories to allow specifying a KMS key. Fixes #1203.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ecs_ecr

##### ADDITIONAL INFORMATION
```paste below

```

### CODE REVIEW

File: plugins/modules/ecs_ecr.py [hunk 1]
```diff
@@ -248,17 +272,21 @@ def get_repository_policy(self, registry_id, name):
...
-    def create_repository(self, registry_id, name, image_tag_mutability):
+    def create_repository(self, registry_id, name, image_tag_mutability, encryption_configuration):
         if registry_id:
             default_registry_id = self.sts.get_caller_identity().get('Account')
             if registry_id != default_registry_id:
                 raise Exception('Cannot create repository in registry {0}.'
                                 'Would be created in {1} instead.'.format(registry_id, default_registry_id))
 
+        if encryption_configuration is None:
+            encryption_configuration = dict(encryptionType='AES256')
```
Comment (markuman) [inline plugins/modules/ecs_ecr.py:283]:
```suggestion
        if encryption_configuration is None:
            encryption_configuration = dict(encryptionType='AES256')
```

This makes it impossible to create a none-encrypted ECR.

Comment (markuman) [reply plugins/modules/ecs_ecr.py:283]: ah wait. AES256 is the default

File: plugins/modules/ecs_ecr.py [hunk 2]
```diff
@@ -85,6 +85,23 @@
...
+                    - The encryption type to use.
+                choices: [AES256, KMS]
+                default: 'AES256'
+                type: str
+            kmsKey:
+                description:
+                    - If I(encryption_type=KMS), specify the KMS key to use for encryption.
+                    - The alias, key ID, or full ARN of the KMS key can be specified.
+                type: str
+        type: dict
```
Comment (markuman) [inline plugins/modules/ecs_ecr.py:104]:
```suggestion
    encryption_configuration:
        description:
            - The encryption configuration for the repository.
        required: false
        suboptions:
            encryptionType:
                description:
                    - The encryption type to use.
                choices: [AES256, KMS]
                default: 'AES256'
                type: str
            kmsKey:
                description:
                    - If I(encryption_type=KMS), specify the KMS key to use for encryption.
                    - The alias, key ID, or full ARN of the KMS key can be specified.
                type: str
        type: dict
        version_added: 5.2.0
```

**Comment by markuman** [comment]
@rwha thanks for your first contribution.  
Can you also please expand the integration test `tests/integration/targets/ecs_ecr/`?

**Comment by rwha** [comment]
I have something written to check for modifications to the encryption config, but I'm unsure of how to handle an unsupported change. 

If the encryption configuration for an existing repository is different from what is defined in the task, should the task fail, or should a warning be emitted? I would prefer a failure, however showing a warning was mentioned in #1203.  If a failure is preferred, should the code at that point do something like this?

```python
result['msg'] = 'Repository encryption configuration cannot be modified.'
return False, result
```

**Comment by markuman** [comment]
> If the encryption configuration for an existing repository is different from what is defined in the task, should the task fail, or should a warning be emitted? I would prefer a failure,

I'll prefer a warning instead of a complete interruption.  
Imo it must use also `module.warn( ... )`

But let's wait was others are expecting/prefer.  
cc @alinabuzachis @jatorcasso @jillr @tremble 

TL;DR: Once the ECR is created, the encryption mode is not changeable. What should the module do when the requested encryption differs from existing encryption? Fail or warn?

**Comment by tremble** [comment]
> TL;DR: Once the ECR is created, the encryption mode is not changeable. What should the module do when the requested encryption differs from existing encryption? Fail or warn?

I would lean towards **fail** rather than *warn*.  Especially with something security related: When Ansible's running within a tool like Ansible ~Tower~ Automation Platform, the warnings may be completely missed unless someone goes explicitly looking for them.  While it seems minor, with the same encryption algorithms in use, the difference between who controls the encryption keys, and what keys are reused for what purposes may be important to folks.

File: plugins/modules/ecs_ecr.py [hunk 3]
```diff
@@ -550,7 +581,8 @@ def main():
         purge_policy=dict(required=False, type='bool'),
         lifecycle_policy=dict(required=False, type='json'),
         purge_lifecycle_policy=dict(required=False, type='bool'),
-        scan_on_push=(dict(required=False, type='bool', default=False))
+        scan_on_push=(dict(required=False, type='bool', default=False)),
+        encryption_configuration=(dict(required=False, type='dict'))
```
[the remark this answers is not in the document]
Comment (tremble) [reply plugins/modules/ecs_ecr.py:585]: Please also define what's supposed to be in this dict, It's relatively easy, just adding in the 'options' dict.  Retroactively applying this can be messy, since you go from the state of garbage being dropped on the floor, to garbage causing errors.

Something like:
```
  encryption_configuration=(
    dict(
      required=False,
      type='dict',
      options=dict(
        encryption_type=dict(required=True, type='str', choices=['AES256', 'KMS'],)
        kms_key=dict(required=False, type='str')
      ),
      required_if=('encryption_type', 'KMS', ['kms_key']),
    ),
  ),
```

Note: you can also add things like dependencies/conflicts between the sub-options.

File: plugins/modules/ecs_ecr.py [hunk 4]
```diff
@@ -85,6 +85,24 @@
         default: false
         type: bool
         version_added: 1.3.0
+    encryption_configuration:
+        description:
+            - The encryption configuration for the repository.
+        required: false
+        suboptions:
+            encryptionType:
```
Comment (tremble) [inline plugins/modules/ecs_ecr.py:93]: It's generally preferred to use snake_case rather than camelCase for option and suboption names.  I know that this is being directly passed to the API, however we try to avoid mixing things up unless absolutely necessary (for example we don't mess with IAM policies where case can be important)

**Review by tremble — CHANGES_REQUESTED** [review]

**Review by markuman — APPROVED** [review]

**Review by alinabuzachis — APPROVED** [review]

File: tests/integration/targets/ecs_ecr/tasks/main.yml [hunk 5]
```diff
@@ -10,6 +10,24 @@
     - set_fact:
         ecr_name: '{{ resource_prefix }}-ecr'
 
+    - name: get ARN of calling user
+      aws_caller_info:
+      register: aws_caller_info
+
+    - name: create KMS key for testing
+      aws_kms:
+        alias: "{{ resource_prefix }}-kms"
```
Comment (tremble) [inline tests/integration/targets/ecs_ecr/tasks/main.yml:19]:
```suggestion
        alias: "{{ resource_prefix }}-ecr"
```

File: tests/integration/targets/ecs_ecr/tasks/main.yml [hunk 6]
```diff
@@ -10,6 +10,24 @@
...
+        description: a key used for testing ECR
+        state: present
+        enabled: yes
+        key_spec: SYMMETRIC_DEFAULT
+        key_usage: ENCRYPT_DECRYPT
+        policy: "{{ lookup('template', 'kms_policy.j2') }}"
+        tags:
+          Name: "{{ resource_prefix }}-kms"
```
Comment (tremble) [inline tests/integration/targets/ecs_ecr/tasks/main.yml:27]:
```suggestion
          Name: "{{ resource_prefix }}-ecr"
```

**Review by tremble — APPROVED** [review]
I'm not overly familiar with ECR, however in general things look good
